[OU-ADD] account: Migration to 16.0

daiduongnguyen-odoo opened the pull request:
- [x] account.accouunt: Account type in account template and account
- [x] account.move: auto_post
- [x] account.move.line: Delete sql_contraints, display_type
- [x] account.move.line: Dynamic analytic_distribution fill yeahhh
- [x] account.journal: Delete sql_contraints
- [x] account.payment: Add field amount_company_currency_signed
- [x] account.bank.statement: internal_index, first_line_index, is_complete

https://youtu.be/f1vUk11DpiU (credit : @ngochung207 )

ngochung207 committed:
[OU-ADD] accont: Migration to 16

account: more analysis on xml
account.accouunt: Account type in account template and account
 - account.move: auto_post
 - account.move.line: Delete sql_contraints, display_type
 - account.move.line: Dynamic analytic_distribution fill yeahhh
 - account.journal: Delete sql_contraints
 - account.payment: Add field amount_company_currency_signed
 - account.bank.statement: internal_index, first_line_index, is_complete

daiduongnguyen-odoo committed:
[OU-IMP] account: simplify analytic_distribution generate

ngochung207 commented:
@duong77476 : add code: `openupgrade.rename_xmlids(env, _xmlids_renames)` in pre-migration

daiduongnguyen-odoo commented:
> 

Yes sir

royle-vietnam commented on openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py:
```diff
@@ -0,0 +1,634 @@
+    analytic.account.id and it has 1 analytic.tag also have that analytic.account then
+    the value will sum together
+    """
+    if not openupgrade.column_exists(env.cr, table_name, "analytic_distribution"):
```
@duong77476 code này khá phức tạp, anh thấy tính tái sử dụng của nó cũng không cao
Do có model thì lấy tài khoản thông qua 1 model khác, hoặc quan hệ của nó khá lằng nhằng
Em viết riêng cho account_move_line thôi nhé, ông nào cần thì copy lại sau
Em tham khảo đoạn code bên dưới nhé
Giải pháp:
- lấy tất cả các move line, cứ có tài khoản KTQT thì là 100%
- hợp thêm với các thẻ KTQT có phân bổ
- sau đó sum chúng lại với nhau theo tài khoản 
```python
WITH distribution_data AS (
    WITH sub AS (
        SELECT
            all_line_data.move_line_id,
            all_line_data.analytic_account_id,
            SUM(all_line_data.percentage) AS percentage
        FROM (
            SELECT
                move_line.id AS move_line_id,
                account.id AS analytic_account_id,
                100 AS percentage
            FROM account_move_line move_line
            JOIN account_analytic_account account
                ON account.id = move_line.analytic_account_id
            WHERE move_line.analytic_account_id IS NOT NULL

            UNION ALL

            SELECT
                move_line.id AS move_line_id,
                dist.account_id AS analytic_account_id,
                dist.percentage AS percentage
            FROM account_move_line move_line
            JOIN account_analytic_tag_account_move_line_rel tag_rel
                ON tag_rel.account_move_line_id = move_line.id
            JOIN account_analytic_distribution dist ON dist.tag_id = tag_rel.account_analytic_tag_id
        ) AS all_line_data
        GROUP BY all_line_data.move_line_id, all_line_data.analytic_account_id
    )
    SELECT sub.move_line_id, jsonb_object_agg(sub.analytic_account_id::text, sub.percentage) AS analytic_distribution
    FROM sub
    GROUP BY sub.move_line_id
)
UPDATE account_move_line move_line SET analytic_distribution = dist.analytic_distribution
FROM distribution_data dist WHERE move_line.id = dist.move_line_id
```

daiduongnguyen-odoo replied:
Dạ vâng để e test lại thử với code mới cc @ngochung207 
@SonCrits A Sơn migrate con `Purchase` xem nè a

daiduongnguyen-odoo replied:
@royleviindoo Em test thử thì ổn áp anh ạ cc @ngochung207 @SonCrits

nguyenvietlam0640 commented:
chỗ này khi migration anh xóa giúp em xmlid `action_bank_statement_form_bank` nữa nhé, ở pre hay post đều được.
Vấn đề:
nếu migrate riêng `account` thì không sao, nhưng khi migrate `viin_account_reconciliation` sẽ raise ra lỗi như sau:
![image](https://github.com/Viindoo/OpenUpgrade/assets/106436992/2d7f543a-c9a4-4a68-a5a8-1c879e1b52fa)

Lý do:
Ở bản odoo 15 vẫn còn `action_bank_statement_form_bank` nhưng trên odoo16 thì không còn nữa. Nếu không xoá xmlid này đi thì ORM sẽ tự xử lí sau cùng. Nhưng trên bản 16 module  `viin_account_reconciliation` chúng ta lại thêm 1 xmlid mới có tên là `action_bank_statement_view_form` có cùng tính chất với action trên thành ra cũ chưa xóa mà mới đã thêm vào nên dính unique constraint.

royle-vietnam commented on openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py:
```diff
@@ -0,0 +1,594 @@
+                )
+            ],
+        )
+    data_query = """SELECT res_id FROM ir_model_data
+        WHERE module='%s' AND name='%s'"""
+    for (account_type_old, account_type_new) in ACCOUNT_TYPE:
+        get_data_query = data_query % tuple(account_type_old.split("."))
+        query = f"""
+            WITH account_type_old AS (
+                SELECT id AS account_id
+                FROM account_account
+                WHERE user_type_id IN (
+                    {get_data_query}
+                )
+            )
+            UPDATE {table}
+            SET account_type = '{account_type_new}'
+            WHERE id IN (SELECT account_id FROM account_type_old)
+        """
+        openupgrade.logged_query(env.cr, query)
```
@duong77476 thay bằng đoạn này giúp anh
SQL đã rất phức tạp rồi, nên hạn chế code động để sau này còn bảo trì được
Ba tháng đến 1 năm sau quay lại, sẽ rất khó đọc
```python
WITH account_type_map AS (
	SELECT
		res_id AS user_type_id,
		CASE
			WHEN name = 'data_account_type_receivable' THEN 'asset_receivable'
			WHEN name = 'data_account_type_payable' THEN 'liability_payable'
			WHEN name = 'data_account_type_liquidity' THEN 'asset_cash'
			WHEN name = 'data_account_type_credit_card' THEN 'liability_credit_card'
			WHEN name = 'data_account_type_current_assets' THEN 'asset_current'
			WHEN name = 'data_account_type_non_current_assets' THEN 'asset_non_current'
			WHEN name = 'data_account_type_prepayments' THEN 'asset_prepayments'
			WHEN name = 'data_account_type_fixed_assets' THEN 'asset_fixed'
			WHEN name = 'data_account_type_current_liabilities' THEN 'liability_current'
			WHEN name = 'data_account_type_non_current_liabilities' THEN 'liability_non_current'
			WHEN name = 'data_account_type_equity' THEN 'equity'
			WHEN name = 'data_unaffected_earnings' THEN 'equity_unaffected'
			WHEN name = 'data_account_type_revenue' THEN 'income'
			WHEN name = 'data_account_type_other_income' THEN 'income_other'
			WHEN name = 'data_account_type_expenses' THEN 'expense'
			WHEN name = 'data_account_type_depreciation' THEN 'expense_depreciation'
			WHEN name = 'data_account_type_direct_costs' THEN 'expense_direct_cost'
			WHEN name = 'data_account_off_sheet' THEN 'off_balance'
		END AS account_type
	FROM ir_model_data
	WHERE module='account' AND name IN (
		'data_account_type_receivable',
		'data_account_type_payable',
		'data_account_type_liquidity',
		'data_account_type_credit_card',
		'data_account_type_current_assets',
		'data_account_type_non_current_assets',
		'data_account_type_prepayments',
		'data_account_type_fixed_assets',
		'data_account_type_current_liabilities',
		'data_account_type_non_current_liabilities',
		'data_account_type_equity',
		'data_unaffected_earnings',
		'data_account_type_revenue',
		'data_account_type_other_income',
		'data_account_type_expenses',
		'data_account_type_depreciation',
		'data_account_type_direct_costs',
		'data_account_off_sheet'
	)
)
UPDATE account_account aa
	SET account_type = atm.account_type
FROM account_type_map atm
WHERE atm.user_type_id = aa.user_type_id
```

daiduongnguyen-odoo replied:
Vâng anh cc @ngochung207

royle-vietnam commented on openupgrade_scripts/scripts/account/16.0.1.2/upgrade_analysis_work.txt:
```diff
@@ -0,0 +1,554 @@
+account      / account.account          / user_type_id (many2one)       : DEL relation: account.account.type, required
+# DONE: pre-migration: _fast_fill_account_account_type
+
+account      / account.account          / include_initial_balance (boolean): NEW isfunction: function, stored
```
@duong77476 logic compute của nó khá đơn giản, nên điền dữ liệu trước cho nó

daiduongnguyen-odoo replied:
Vâng a

royle-vietnam commented on openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py:
```diff
@@ -0,0 +1,594 @@
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_analytic_line AS aal
+        SET journal_id = (
+            SELECT journal_id
+                FROM account_move_line AS aml
+                WHERE aml.id = aal.move_line_id
+            LIMIT 1
```
@duong77476 như này cho đơn giản nhé
```suggestion
        UPDATE account_analytic_line aal
            SET journal_id = aml.journal_id
        FROM account_move_line aml
        WHERE aml.id = aal.move_line_id
```

daiduongnguyen-odoo replied:
Đoạn này trc e code quên chưa sửa, e sẽ sửa lại

daiduongnguyen-odoo commented:
> action_bank_statement_form_bank

Chưa chắc chuẩn mà xử ở account cũng ko chuẩn bằng xử ở `viin_account_reconciliation,` a xử ở đây rồi  https://github.com/Viindoo/odoo-openupgrade-tvtmaaddons/pull/428

royle-vietnam commented on openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py:
```diff
@@ -0,0 +1,594 @@
+                JOIN tmp
+                    ON tmp.tag_id = aat.id
+            GROUP BY
+                aad1.analytic_id,
+                aad1.partner_id,
+                aad1.product_id,
+                aad1.company_id,
+                aad1.create_date,
+                aad1.write_date,
+                aad1.create_uid,
+                aad1.write_uid
+        )
+        INSERT INTO account_analytic_distribution_model
+        (partner_id,
+        product_id,
+        company_id,
+        create_date,
+        write_date,
+        create_uid,
+        write_uid,
+        analytic_distribution)
+        SELECT
+            partner_id,
+            product_id,
+            company_id,
+            create_date,
+            write_date,
+            create_uid,
+            write_uid,
+            jsonb_object_agg(analytic_account_of_tag::text, percentage) AS analytic_distribution
+        FROM
+            sub
+        GROUP BY
+            partner_id,
+            product_id,
+            company_id,
+            create_date,
+            write_date,
+            create_uid,
+            write_uid;
```
@duong77476 thay bằng đoạn này giúp anh nhé
```sql
WITH distribution_data AS (
    WITH sub AS (
        SELECT
            all_line_data.analytic_default_id,
            all_line_data.analytic_account_id,
            SUM(all_line_data.percentage) AS percentage
        FROM (
            SELECT
                analytic_default.id AS analytic_default_id,
                account.id AS analytic_account_id,
                100 AS percentage
            FROM account_analytic_default analytic_default
            JOIN account_analytic_account account
                ON account.id = analytic_default.analytic_id
            WHERE analytic_default.analytic_id IS NOT NULL
            UNION ALL
            SELECT
                analytic_default.id AS analytic_default_id,
                dist.account_id AS analytic_account_id,
                dist.percentage AS percentage
            FROM account_analytic_default analytic_default
            JOIN account_analytic_default_account_analytic_tag_rel tag_rel
                ON tag_rel.account_analytic_default_id = analytic_default.id
            JOIN account_analytic_distribution dist
                ON dist.tag_id = tag_rel.account_analytic_tag_id
        ) AS all_line_data
        GROUP BY all_line_data.analytic_default_id, all_line_data.analytic_account_id
    )
    SELECT sub.analytic_default_id AS analytic_default_id,
     jsonb_object_agg(sub.analytic_account_id::text, sub.percentage)
         AS analytic_distribution
    FROM sub
    GROUP BY sub.analytic_default_id
)
INSERT INTO account_analytic_distribution_model (
	account_id,
	partner_id,
	product_id,
	company_id,
	create_date,
	write_date,
	create_uid,
	write_uid,
	analytic_distribution)
SELECT
	aa.code,
	aad.partner_id,
	aad.product_id,
	aad.company_id,
	aad.create_date,
	aad.write_date,
	aad.create_uid,
	aad.write_uid,
	dist.analytic_distribution
FROM
	distribution_data dist
JOIN account_analytic_default aad ON aad.id = dist.analytic_default_id
JOIN account_account aa ON aa.id = aad.account_id
```

daiduongnguyen-odoo replied:
Dạ vâng

daiduongnguyen-odoo replied:
Ý anh là `account_prefix` vì `account_id` không tồn tại trong `account_analytic_distribution_model`, mà chắc là `account_prefix` rồi

royle-vietnam commented on openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py:
```diff
@@ -0,0 +1,594 @@
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_bank_statement AS stmt
+        SET first_line_index = (
+            SELECT internal_index
+            FROM account_bank_statement_line
+            WHERE statement_id = stmt.id
+            ORDER BY statement_id
+            LIMIT 1
```
@duong77476 đoạn sql này có vẻ không đúng với logic compute
cần sắp xếp theo `internal_index` và lấy bản ghi đầu tiên, chứ không phải sắp xếp theo statement_id
ngoài ra, nếu đã compute `first_line_index`, thì cần compute cả `date`, nếu không thì nó cũng sẽ vào compute
tốt nhất anh nghĩ không cần điền giá trị cho trường này, vì cũng không nhiều bản ghi ở model này

daiduongnguyen-odoo replied:
Vâng vậy e sẽ remove luôn thằng  `is_complete` nữa ạ

royle-vietnam commented on openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py:
```diff
@@ -0,0 +1,594 @@
+    openupgrade.logged_query(
+        env.cr,
+        """
+        UPDATE account_bank_statement AS stmt
+        SET is_complete = (
+            CASE
+                WHEN EXISTS (
+                    SELECT 1
+                    FROM account_bank_statement_line
+                    WHERE statement_id = stmt.id
+                    AND stmt.state = 'posted'
+                    AND stmt.balance_end = stmt.balance_end_real
+                    LIMIT 1
+                ) THEN true
+            ELSE false
+        END
```
@duong77476 để orm xử lý nhé
vì so sánh kiểu `balance_end = balance_end_real` sẽ không chuẩn, vì trong code đang so sánh theo tiền tệ nữa
ví dụ nếu là tiền USD, có thể trong database lưu là `balance_end = 0.051 và balance_end_real = 0.053` thì vẫn là True, do tiền USD chỉ so sánh 2 chữ số sau dấu phẩy

daiduongnguyen-odoo replied:
Vâng ạ

royle-vietnam commented on openupgrade_scripts/scripts/account/16.0.1.2/upgrade_analysis_work.txt:
```diff
@@ -0,0 +1,554 @@
+account      / account.reconcile.model.line / analytic_distribution_search (json): NEW
+account      / account.reconcile.model.line / analytic_precision (integer)  : NEW hasdefault: default
+account      / account.reconcile.model.line / analytic_tag_ids (many2many)  : DEL relation: account.analytic.tag
+# TODO: like the change in account.move
```
@duong77476 xử nó giúp anh nhé

daiduongnguyen-odoo replied:
Done ạ

royle-vietnam commented on openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py:
```diff
@@ -0,0 +1,594 @@
+                )
+            ],
+        )
+    # TODO: consider fill value
```
@duong77476 vẫn còn todo à?
trường compute thì có thể nothing todo, để orm tự xử, trước mắt chấp chận thế để đi cho nhanh
nếu sau này chậm thì quay lại dùng sql để xử sau

daiduongnguyen-odoo replied:
Vâng cái này trước em thấy logic compute nó hơi phức tạp nên đánh tạm todo ạ

royle-vietnam commented on openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py:
```diff
@@ -0,0 +1,594 @@
+    Case 5: with aml is an accounting transaction occurring
+            set display type is 'product'
+    """
+    openupgrade.logged_query(
```
@duong77476 viết code như này khó đọc quá, thay cho anh bằng đoạn này nhé
```sql
UPDATE account_move_line aml
    SET display_type = CASE 
		WHEN am.move_type NOT IN ('out_invoice', 'out_refund', 'in_invoice', 'in_refund') THEN 'product'
		WHEN aml.tax_line_id IS NOT NULL THEN 'tax'
		WHEN aa.account_type IN ('asset_receivable', 'liability_payable') THEN 'payment_term'
		ELSE 'product'
	END
FROM account_move am
LEFT JOIN account_account aa ON aa.id = aml.account_id
WHERE aml.display_type IS NULL AND am.id = aml.move_id
```

daiduongnguyen-odoo replied:
@royleviindoo Dạ vâng a, cái này e tập trung xử nốt cái ticket cho khách hàng rồi e quay lại xử liền ạ
cc @ngochung207

daiduongnguyen-odoo replied:
Chỗ này e phải dùng subquery không nó báo kiểu này
![Screenshot from 2023-07-24 21-34-51](https://github.com/Viindoo/OpenUpgrade/assets/56789189/a5594726-81db-4195-b59f-815b20d89b25)

royle-vietnam reviewed (commented):
Sửa hết comment là có thể merge được

daiduongnguyen-odoo committed:
[OU-IMP] account: simplify some fast fill stuff and remove unneeded
method

-display_type improvement
-remove bank statement fastfill
-add account_reconcile_model_line analytic_distribution generate
-improve analytic_distribution_model generation from analytic default

royle-vietnam commented on openupgrade_scripts/scripts/account/16.0.1.2/pre-migration.py:
```diff
@@ -0,0 +1,561 @@
+        UPDATE account_move_line AS aml
+        SET display_type = subquery.display_type
+        FROM (
+            SELECT
+                aml.id,
+                CASE
+                    WHEN am.move_type NOT IN
+                    ('out_invoice', 'out_refund', 'in_invoice', 'in_refund') THEN 'product'
+                    WHEN aml.tax_line_id IS NOT NULL THEN 'tax'
+                    WHEN aa.account_type IN
+                    ('asset_receivable', 'liability_payable') THEN 'payment_term'
+                    ELSE 'product'
+                END AS display_type
+            FROM account_move_line AS aml
+            LEFT JOIN account_move AS am ON am.id = aml.move_id
+            LEFT JOIN account_account AS aa ON aa.id = aml.account_id
+            WHERE aml.display_type IS NULL AND am.id = aml.move_id
```
@duong77476 đưa đống này vào `WITH` nhé em
Mỗi lần update lại select như này sẽ rất chậm

daiduongnguyen-odoo replied:
Dạ vâng
